Document example of Proxy use

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50583 Document example of Proxy use**

Differential Revision: [D26010501](https://our.internmc.facebook.com/intern/diff/D26010501)

Commit:
Document example of Proxy use

[ghstack-poisoned]

Commit:
Update on "Document example of Proxy use"




[ghstack-poisoned]

jamesr66a commented on torch/fx/examples/create_graph_using_proxy.py:
```diff
@@ -0,0 +1,61 @@
+
+# Wrap our created Graph in a GraphModule to get a final, runnable
+# `nn.Module` instance
+mod = GraphModule(Empty(), graph)
```
Instead of defining `class Empty`, can just pass `torch.nn.Module()` here

jamesr66a reviewed (commented):
Looks good, though I think this emphasizes a "boundary case" I guess I would call it, where you're tracing an entire function. It might be good to show also show using Proxy's to inline a given function into another function, or to show using Proxies within some metaprogramming code to emit operations (i.e. not just tracing through straightline code)

Commit:
Update on "Document example of Proxy use"




[ghstack-poisoned]

jamesr66a commented on torch/fx/examples/inline_function.py:
```diff
@@ -0,0 +1,68 @@
+'''
+How to inline a function into an existing Graph
+
+If you're using the default Tracer to symbolically trace, ``torch.nn``
```
Probably want to make it clear this is only an example scenario. The code is good for any situation in which one wants to inline a call

jamesr66a commented on torch/fx/examples/inline_function.py:
```diff
@@ -0,0 +1,68 @@
+            # creating a "closure" that includes those Nodes. The result
+            # of this call is another Proxy, which we can hook into
+            # our existing Graph to complete the function inlining.
+            proxy_output = traced_relu(*proxy_args, **proxy_kwargs)
```
nit: can also just call `m.relu` directly, I realized that after sending you that notebook

jamesr66a commented on torch/fx/examples/inline_function.py:
```diff
@@ -0,0 +1,68 @@
+            # Call `traced_relu` with the newly-created Proxy arguments.
+            # `traced_relu` is the generic version of the function; by
+            # calling it with Proxies created from Nodes in `m`, we're
+            # creating a "closure" that includes those Nodes. The result
```
Use of the term "closure" seems strange to me here (i.e. i can't reconcile it with [this](https://en.wikipedia.org/wiki/Closure_(computer_programming)) definition). In particular, I don't think we're creating a function with an associated environment, but rather emitting nodes potentially referencing existent values in the IR

Maybe I'm just not understanding?

jamesr66a commented on torch/fx/examples/wrap_output_dynamically.py:
```diff
@@ -0,0 +1,73 @@
+def wrap_in_activation_function(m: GraphModule, fn: ActivationFunction) -> GraphModule:
+    # Get output node
+    output_node = next(iter(reversed(m.graph.nodes)))
+    assert output_node.op == "output"
```
Instead of asserting that nodes[-1] is the output node, should probably iterated on `reversed(m.graph.nodes)` and assign to `output` node to an optional name, then assert that that's not none. Only a technicality, but `output` nodes don't necessarily need to be last (any code after an output node is not executed, but w/e)

jamesr66a reviewed (approved):
Looks good, just some small stuff then we should be good to go

Commit:
Update on "Document example of Proxy use"




[ghstack-poisoned]

Commit:
Update on "Document example of Proxy use"


Differential Revision: [D26010501](https://our.internmc.facebook.com/intern/diff/D26010501)

[ghstack-poisoned]
